### ISSUE TITLE
Feature request M109/M190 interruptable

### ISSUE THREAD

**Issue by justmyopinion** [MarlinFirmware/Marlin#685]
It would be nice if heater commands set/temp/wait´like M109/M190 could be interruptable.
If wrong parameters sent or hw errors,  Marlin would hang infinitely.
Somehow it would be nice if these commands could be cancelled by host without a  hw reset.

**Comment by gege2b** [comment]
Hi all,
I can confirm this problem is still here with the latest commit ( 08c236059efc90080d2644a6bc6c882a07b03f6b )

In fact, the problem is when we start a print with a gcode that uses M109/M190 to set the temperature to reach the target.
I think its a good thing : first we set the bed temperature, and when this temperature is reached, we set the extruder temp.

But, if for some reason, the print is canceled (i.e using the "Off" button of printrun) _BEFORE_ the temperature has reached the target, the temperature still continue to rise.

The only way to interrupt this behavior is to reset the printer (hard or soft)

This is why, I too think that M109/190 should be made interruptable (i.e. by any other M109/190 command)

Oh, and I cannot find any implementation of M116 in Marlin, so  I don't know if this solution is effective right now

**Comment by galexander1** [comment]
I don't understand.  M109 sets the target temperature exactly how M104
does, then it waits until the target is reached.  If you interrupt it,
the same target will remain, it will just stop before it is reached -- in
effect you will have changed the M109 into an M104.  If you start a new
print while it is heating, and that new print issues M109, the new target
will be set and it will wait until this new target is reached.

What should be different?

**Comment by justmyopinion** [comment]
@galexander1 
Point is that this behaveour is bad design. Firmware should NEVER be locked up by any command.
It would be far better if HOST is sending M104 to the FW and then polls the temperature by sending M105 for reading of current temperature, regulation in FW is taking care of temperature set. When temperature is reached, print can be continued by host. This assures that any print program can be cancelled by Host any time and buffer is probably empty so temperature settings could easily be reset.

**Comment by galexander1** [comment]
I still don't understand.  I could see why you might want the host to
send M104 and then poll with M105 -- what would need to change in Marlin
for this to happen?

My host sends a reset signal (DTR low or something like that) every time
I start it.  Marlin stops blocking (in fact resets temperature to 0) when
this signal is received.  And I don't use an LCD.

Is the change you are suggesting that the LCD "stop" function should
interrupt M109?

**Comment by justmyopinion** [comment]
No change to Marlin/Repetier etc is needed!. 
M109/190 should never be issued as Host should do the" waiting for", but it would of course need some changes to "Pronterface","Repetier Host" and any other that uses FW (Marlin/Repetier etc) as starting a print needs some different action , but gain is that any started program could always be cancelled in (Host?/Client?) as FW knows nothing about started programs, but is a dummy FW that only executes GCODES sent to it.

**Comment by gege2b** [comment]
@galexander1 it's not how it actually works...

if a M190 S70 is issued in the start of a gcode, a new gcode changing the bed temp would be processed only after the first target is reached... It's not a problem if the second M190 sets a higher target temperature, but it's annoying if it sets temperature to, say, zero. In this case, only a reset would effectively sets the target temp to zero

(I talk about M190 because it's the only one I've tested to be honest, but M109 should behave the same)

**Comment by galexander1** [comment]
@gege2b From my perspective, if I want to interrupt M109/M190, I reset
the robot using the DTR high signal that most hosts send at the beginning
of a print anyways.  So from my perspective, it is already interruptible.

Are you suggesting some other operation should be able to interrupt it?
Do you want the LCD stop button to interrupt it?

**Comment by thinkyhead** [comment]
It does make a certain sense that an M109/M190 or other temperature setting command should be able to supersede a previous M109/M190, with priority if it comes from the host / console. This is a more "soft control" that host software and Pronterface console users can take advantage of – slightly less "drastic" than resetting the robot ~~(which loses your position awareness)~~.

But from a gcode file, a temperature command should only interrupt another temperature command if it's the very next command (or if no movement commands are issued beforehand).

A proposal.

**Comment by whosawhatsis** [comment]
@gege2b M109/190 don't wait for a temperature drop unless you use an R parameter instead of S.

@galexander1 AT90USB boards (like printrboard) don't have a DTR line (because there's no actual serial connection), so they can't auto-reset. Some other boards have/can have/must have auto reset disabled. All the instructions used to say that in order to use RAMPS, you need to cut the autoreset trace. The Azteeg X3 has a jumper for that purpose.

I'd love to have an Mcode to cancel an M109/190 wait, but it's harder than it sounds. The code interpreter fills a buffer with commands while it's waiting, and stops accepting them until space is free in that buffer, so it wouldn't even read the code until the M109/190 finished. It would take an out-of-channel signal (like pulling DTR high) to get around this.

Having the host reinterpret those codes might be a good solution. There are some edge cases where it might be a problem, and you'd have to implement TEMP_RESIDENCY_TIME, TEMP_HYSTERESIS and, TEMP_WINDOW in the host without access to the firmware's values, though you might be able to wait until the temp hits the target and THEN send an M109/190 to get around this.

In any case, there's nothing to do in Marlin to implement this solution. It would be entirely up to the host software.

**Comment by gege2b** [comment]
@galexander1 it's exactly what I suggest, being able to interrupt M109/190 from another temperature command
I'm not using LCD, I do all my prints from my computer

My though is that if a user wishes to cancel a print (without reseting the printer), all operations should be interrupted, even the temp. rising when the cancel is done while the printer waits for a M109/190 to reach the target temp.

I agree with @thinkyhead proposal

@whosawhatsis well, I missed this, but it doesn't change the whole point anyway

**Comment by thinkyhead** [comment]
Actually, I think the current (newfangled) code that handles multi-line g-code comes close to having a "front-buffer." (Or maybe that's yet to be introduced.) Commands in a _front buffer_ will go through ahead of others, even if things are backed up in the normal-priority buffer. The front-buffer is necessarily short, and only immediate commands (changing modes is ok) should go in. Then you just have a gatekeeper (function) that decides if commands in the input g-code stream (sd, usb) go into one buffer or the other.

There are other (not-so-crazy) approaches that could be helpful and improve user experience. I'm thinking in particular that since only extruder moves are prohibited by a low extruder temperature, the robot doesn't have to pause altogether. Instead of having temperature commands that wait and block (actually blocking execution), they should instead just change the "machine state." An if-block in the main loop then checks the temperature in this state, while some other activities are allowed, including _immediate_ commands (to the front buffer, if they pass).

When temperature is reached, the callback that was specified in the original M109/190 command gets fired. In the meantime, the sd-card or usb printing proceeds, with g-code that moves XYZ allowed through, but movement of E axes is deferred in the buffer until temperature is reached. That's the point where blocking happens. I foresee one (big?) problem with this "modal" approach: the nozzle will most likely end up sitting right on the bed while awaiting temperature to be reached....

Getting too creative maybe, but something about a "finite-state-machine" approach always has had appeal.

**Comment by daid** [comment]
I think this should simply be handled in the host side, not on the printer side. Else you'll will be generating a huge mess. (And, it's actually not that hard to handle host side)

(Thus, sending M104/M140 instead of M109/M190 and monitor the temperature from the host till the target is reached)

**Comment by gege2b** [comment]
Well, I'll try to never have to abort a print mid M190 again then.

**Comment by nophead** [comment]
How does that fix the problem when printing from SD but started with a
host? The SD file has to have M190 but once started you can't stop it if
you change your mind.

**Comment by thinkyhead** [comment]
@gege2b @nophead Those are "common use cases." Wot to do, wot to do?

**Comment by daid** [comment]
I cancel the M109 and M190 by setting the target temperatures to 0 on an abort in the UM2. (it has no wait on cooldown, but you could add a special case for a target temperature of 0 to work for all cases)

**Comment by scotty1024** [comment]
Reset doesn't clear SRAM but powering off/on does. If you invest two bytes as "soft reset sentinels" you can determine the type of reset that just occurred. This lets operators use the reset button on the LCD/mainboard to interrupt operations yet not lose printer awareness.

**Comment by justmyopinion** [comment]
@daid 
Would you be the first to implement this waiting_for_temp functionality in HOST sw and stop using M109/190 sent to printer? and next make wait function cancellable by user? 
I would love to see this addition :-) and would be happy to challenge other host sw with this new functionality.
It does not necessary mean that current functionality would become obsolete.
As  @whosawhatsis (try to pronounce that) has mentioned it should not be a big deal to add this functionality to host sw.

**Comment by thinkyhead** [comment]
In my fantasy _front queue_ a Cancel command received from anywhere would go to the front. :hand:

**Comment by nophead** [comment]
Can't M190 just terminate as soon as something arrives in the buffer as the
host shouldn't send anything while it waits for M190 to finish and come
back with ok.

**Comment by daid** [comment]
@nophead nope, as you send a few lines "in advance" from the host, to keep buffers filled and lower the effect of the large round-trip times of USB-Serial.

**Comment by nophead** [comment]
Well the host could desist from filling the buffer when it sees M190 as it
is pretty pointless.

**Comment by daid** [comment]
Well, yes, but if you are making a special exception for M190/M109 anyhow, you better just handle it all host site as I suggested before. Which is even backwards compatible.

**Comment by nophead** [comment]
But doesn't work for SD prints.

**Comment by nophead** [comment]
BTW, the host and firmware should exchange protocol version numbers to
allow changes like this without compatibility issues.

**Comment by justmyopinion** [comment]
@nophead 
well...a lot of things are different with SD prints, maybe that is why M109/190 should survive in FW, but with host it would be better to take control when these commands are detected and send m104 commands to fw and handle the waiting in host sw.

I do not see any compatibility issues if host sw just handle m109/190 commands transparent.

**Comment by nophead** [comment]
I
Yes but I run SD prints from a host and sometimes want to cancel it

**Comment by gege2b** [comment]
looking at the code, there is a boolean named "cancel_heatup" in the M190 section, it does exactly what it's supposed to do, but only when using a panel/LCD via a M0 or M1

is it possible to set this bool via something like M140/M104 S0 too while in the heatup cycle ?

I admit I don't perfectly understand how buffer works on marlin, thats why my question

**Comment by thinkyhead** [comment]
@gege2b Sounds like currently any command that could cancel heating would be stuck in the buffer until heating finishes. Need some way to get past that...

**Comment by cscott** [comment]
This bug should be reopened, it's definitely not fixed yet.

**Comment by daid** [comment]
It's not fixable.

**Comment by scotty1024** [comment]
Daid,

Suggestion for host software fix for this issue.

As commands are parsed from the g-code file inspect them to see if they are one of the set temperature and wait commands. If one is found re-write it as a set temperature command.

Host software then polls waiting for temperature to be reached. No other g-code statements are sent from the file so interrupting the heating cycle is very simple.

People that wish to be able to easily cancel warm up's will appreciate this feature (which you can make selectable on/off for those that don't wish it) will see this as another reason to use your tool.
